src/ConstantInterval.cpp

```diff
@@ -584,36 +588,51 @@ ConstantInterval abs(const ConstantInterval &a) {
 }
 
 ConstantInterval operator<<(const ConstantInterval &a, const ConstantInterval &b) {
-    // Try to map this to a multiplication and a division
-    ConstantInterval mul, div;
-    constexpr int64_t one = 1;
-    if (b.min_defined) {
-        if (b.min >= 0 && b.min < 63) {
-            mul.min = one << b.min;
-            mul.min_defined = true;
-            div.max = one;
-            div.max_defined = true;
-        } else if (b.min > -63 && b.min <= 0) {
-            mul.min = one;
-            mul.min_defined = true;
-            div.max = one << (-b.min);
-            div.max_defined = true;
+    // In infinite integers (with no overflow):
+
+    // a << b == a * 2^b
+
+    // This can't be used directly, because if b is negative then 2^b is not an
+    // integer. Instead, we'll break b into a difference of two positive values:
+    // b = b_pos - b_neg
+    // So
+    // a * 2^b
+    // = a * 2^(b_pos - b_neg)
+    // = (a * 2^b_pos) / 2^b_neg
+
+    // From there we can use the * and / operators.
+
+    ConstantInterval b_pos = max(b, 0), b_neg = max(-b, 0);
+
+    // At this point, we have sliced the interval b into two parts. E.g.
+    // if b = [10, 12],  b_pos = [10, 12] and b_neg = [0, 0]
+    // if b = [-4, 8],   b_pos = [0, 8]   and b_neg = [0, 4]
+    // if b = [-10, -3], b_pos = [0, 0]   and b_neg = [3, 10]
+    // if b = [-3, inf], b_pos = [0, inf] and b_neg = [0, 3]
+    // In all cases, note that b_pos - b_neg = b by our definition of - for
```

Comment:
I think this is not true if b.min is defined and equal to INT64_MIN, because `(b_pos - b_neg).min` is not defined.

Reply:
@abadams does this matter? If not, this PR lgtm

Reply:
I had to think about it, so it warrants updating the comment. I've added: 

(ignoring corner cases, for which b_pos - b_neg safely over-approximates the bounds of b).